# Introduction to Python for Statistical Data Analysis

**Introduction**

Start by discovering some Python fundamental concepts. 



**Variables and Data Types**

Learn about Python data types: integer, float, and string.

**Basic Input and Output**

     - Output: print()
     - Input: input()

**Data Structures**

Learn how to store, access, and manipulate data in Python data structures.

    - List
    - Tuple
    - Exercise: working with lists & tuples
    - Dictionary
    - Exercise: working with dictionaries

**Conditional and Iterative Processing**

Learn about different comparison operators, how to combine them with Boolean operators, and how to use the Boolean outcomes in control structures. 

     - if, elif, else
         - Exercise: working with conditional statements
     - Loops
         - for loops
         - while loops
         - Exercise: working with loops

**Functions**

Learn how to use functions and how to create your own.


Exercise: working with functions
  

**Reading & Writing Data Files**

Learn the many ways in which to import data into Python: from flat files such as .txt and .csv to other files from other software such as Excel spreadsheets, Stata, SAS, and MATLAB files.

     - Importing text files
     - Importing flat files
         - Importing flat files with pandas
     - Importing other file types
         - Importing Excel spreadsheets
         - Importing SAS, Stata and MATLAB files
     - Writing files 

**NumPy** 

*NumPy* is a fundamental Python package to efficiently practice data science. Learn to work with powerful tools in the NumPy array!

     - NumPy arrays
     - 2D NumPy arrays
     - Basic statistics with NumPy
     - Exercise: working with NumPy

***pandas*** 

Learn about the pandas *DataFrame*, the de facto standard to work with tabular data in Python. Practice creating and manipulating datasets, and learn how to access the information you need from these data structures.

     - pandas data structures
         - Series
         - DataFrame
         - Exercise: working with pandas DataFrame
     - Manipulating DataFrames with pandas
         - Filtering DataFrames
         - Merging DataFrames
         - Reshaping DataFrames
         - Grouping DataFrames
     - Statistics with pandas
     - Exercise: data manipulation with pandas

**Data Visualization with *matplotlib* & *Seaborn***

Learn how to build various types of plots with *matplotlib* and *Seaborn*. Then focus on how customize them to be more visually appealing and interpretable.

         - matplotlib
         - Plotting with seaborn & pandas
             - Visualizing two quantitative variables
             - Visualizing categorical variables
             - Extra features for costumizing plots in Seaborn 

**Introduction to Statistical Analysis**

An introduction to the basics of statistics with Python, using the stats functions in the *scipy* and *statsmodels* Python packages.

         - Two-sample t-test
             - Assumption checks
             - Two samples t-test
         - Chi square test
             - Constructing the contingency table
             - Calculating the Chi square test
         - Exercise 1
         - Simple linear regression

**Final Exercise**

# Introduction

Python is an interpreted, high-level, general-purpose programming language. Created by Guido van Rossum and first released in 1991, Python's design philosophy emphasizes code readability with its notable use of significant whitespace. Its language constructs and object-oriented approach aim to help programmers write clear, logical code for small and large-scale projects.

In Python, everything is an object. Every number, string, data structure, function, ... exists in its own container which is referred to as Python object. Each object has an associated type and internal data. 

The single use of the equality sign (=) refers to the **assignment operator**.

In [8]:
x = 3 + 9
x

12

In [10]:
y = 5 * 2
y

10

Assigment is also referred to as binding, as we are binding a name to an object.  

The double equality sign (==) refers to the actual equality operator.

In [7]:
x = 3 + 5
x == 8

True

Inequality is expressed as != . Using the same example as before, we get the following: 

In [12]:
x = 3 + 5
outcome = x != 9

In [14]:
outcome # this is a comment

True

The results of equality and inequality statements are boolean, meaning True or False. 

In [16]:
print ( " Hello , Actuarial Science students ! " )
print ( " Welcome to Python . " )
print ( " Let us begin ! " )

 Hello , Actuarial Science students ! 
 Welcome to Python . 
 Let us begin ! 


Remember that:
* Object names in Python are **case-sensitive**. This means that Gender and gender refer to different objects. 
* To add **comments** to the code, you can use the hash mark # and the text following it will be ignored by Python and not executed. 
    
**Where to ask for help?**

The Python help function is used to display the documentation of modules, functions, classes, keywords, ...
The help function has the following syntax:

In [ ]:
help(print)

#  Variables and Data Types

## Creating variables

In Python, you create (declare) a variable simply by assigning a value to a name using the = sign:

In [24]:
name = "Alice"
age = 20
height = 1.75
is_student = True

In [26]:
print(name)
print(age)
print(height)
print(is_student)

Alice
20
1.75
True


## Data Types

There are four key data types in Python:
* numeric types: integer (int) and float.
  * integer (int): Whole numbers (integers)
  * float : Decimal numbers
* text types: string (str)- Text (string of characters) 
* boolean/logical types: boolean (bool)- True or False values

An **integer** is a number without a decimal point.

In [28]:
15

15

You can check the type of an object using the type method.

In [30]:
type(15)

int

In [32]:
a = 10//3 # floor/integer division -  This divides and rounds down to the nearest whole number
print(a)
print(type(a))

3
<class 'int'>


In [34]:
b = 10 % 3 # modulus ---remainder after division
print(b)
print(type(b))

1
<class 'int'>


A **float** is a floating point number, which represents a decimal number.

In [36]:
15.6

15.6

In [38]:
type(15.6)

float

In [40]:
c = 10/3
print(c)
print(type(c))

3.3333333333333335
<class 'float'>


**Text** is stored in a string, a sequence of characters.

In [43]:
"15.0 is a number"

'15.0 is a number'

In [45]:
type("15.0 is a number")

str

Logicals or booleans are stored in Python in objects of the **bool** type. Their only possible values are True and False.

In [47]:
example = False
example

False

In [49]:
type(example)

bool

In [51]:
x = 2
y = 3
x + y<16

True

In [53]:
type(x + y<16)

bool

## Basic Input and Output

### Output: print()

We have already seen the print() function. You can print multiple items by separating them with commas:

In [57]:
name = "Alice"
age = 20
print("Name:", name, "Age:", age)
print(name,age)

Name: Alice Age: 20
Alice 20


You can also use f-strings (formatted strings) for cleaner output:

In [63]:
name = "Alice"
age = 20 +4 -6 
print(f"My name is {name} and I am {age} years old.")

My name is Alice and I am 18 years old.


### Input: input()

The input() function allows you to get information from the user. Whatever the user types is stored as a string.

In [65]:
name = input("Enter your name: ")
print(f"Hello, {name}!")

Enter your name:  Beryl


Hello, Beryl!


## Arithmetic Operators

Python can perform mathematical calculations. The basic arithmetic operators are shown

In [10]:
import pandas as pd

In [12]:
# Create the table data
data = {
    "Operator": ["+", "-", "*", "/", "//", "%", "**"],
    "Description": [
        "Addition",
        "Subtraction",
        "Multiplication",
        "Division",
        "Floor Division",
        "Modulus (remainder)",
        "Exponentiation"
    ],
    "Example": ["5 + 3", "5 - 3", "5 * 3", "5 / 2", "5 // 2", "5 % 2", "2 ** 3"],
    "Result": [8, 2, 15, 2.5, 2, 1, 8]
}

In [14]:
# Create DataFrame
df = pd.DataFrame(data)

# Display table
df

,Operator,Description,Example,Result
0,+,Addition,5 + 3,8.0
1,-,Subtraction,5 - 3,2.0
2,*,Multiplication,5 * 3,15.0
3,/,Division,5 / 2,2.5
4,//,Floor Division,5 // 2,2.0
5,%,Modulus (remainder),5 % 2,1.0
6,**,Exponentiation,2 ** 3,8.0


In [1]:
a = 10
b = 3

In [3]:
print (a + b) # 13

13


In [5]:
print (a - b) # 7

7


In [7]:
print (a * b) # 30

30


In [9]:
print (a / b) # 3.3333...

3.3333333333333335


In [11]:
print (a // b) # 3 ( drops the decimal )

3


In [13]:
print (a % b) # 1 ( remainder of 10 / 3)

1


In [15]:
print (a ** b) # 1000 (10 to the power of 3)

1000


# Data Structures 

In this chapter, we will focus on Python data structures: lists, tuples and dicts. Data structures in Python are simple but powerful.

## Lists

Lists allows you to store objects of different types (integer, float, string, ...). Lists are *mutable*, so their content can be modified *in-place*. List can be created using square brackets [ ] or using the list type function.

In [19]:
fruits = [" apple ", " banana ", " mango ", " orange ", " grape "]
fruits

[' apple ', ' banana ', ' mango ', ' orange ', ' grape ']

In [35]:
list_fruits = list(fruits)

In [21]:
print ( fruits [0]) # apple ( first item , index 0)

 apple 


In [23]:
print ( fruits [2]) # mango ( third item , index 2)

 mango 


In [25]:
print ( fruits [ -1]) # grape ( last item )

 grape 


In [27]:
print (len( fruits )) # 5 ( number of items )

5


In [29]:
first_list = [2, 4, 6, 8, 10]
first_list

[2, 4, 6, 8, 10]

In [31]:
words = ('hello', 'ciao', 'hola')
list_words = list(words)
list_words

['hello', 'ciao', 'hola']

In [33]:
mix_list = ['hello', 2, 'hola', 8]
mix_list

['hello', 2, 'hola', 8]

In [35]:
type(list_words) # check the type

list

In [37]:
type(words) # check the type

tuple

List creation is easy. However, you also need to access a list. In Python, sequences are 0-indexed which means to access the first element of the list you need to ask for: 

In [39]:
mix_list[2]

'hola'

Check the type of the first item and second in the mix_list object below.

In [42]:
type(mix_list[0])

str

In [44]:
type(mix_list[1])

int

A negative index **accesses** items from the end of the list (counting backwards):

In [65]:
mix_list[-1]

'hello'

Pay attention: accessing an item of a list does not change the list.

In [ ]:
mix_list

## Modifying Lists

Elements can be added or removed from a list. To add an element at the end of a list you can use the *append* method.

In [46]:
numbers = [3, 1, 4, 1, 5, 9, 2]
numbers . append (6) # add 6 to the end
numbers . remove (1) # remove first occurrence of 1
numbers . sort () # sort in ascending order
print ( numbers )

[1, 2, 3, 4, 5, 6, 9]


append(mix_list, options) # function

mix_list.append(options) # method

In [48]:
mix_list.append('salut')
mix_list

['hello', 2, 'hola', 8, 'salut']

Append, like most of the methods for lists we present here, works *in-place*, meaning that it does not perform the operation on a copy of the list, it actually changes the original list! The help page for list.append does not mention making a copy of the list.

In [ ]:
help(list.append)

To remove the **last item** in a list, you can use the *pop* method. It actually returns the last element, and then removes it.

In [50]:
mix_list.pop()

'salut'

In [52]:
mix_list

['hello', 2, 'hola', 8]

What happens when you provide the pop methods with an argument?

In [ ]:
new_list = [2, 3, 'four']
new_list.pop(1)
new_list

With *insert* you can add an element in a specific location in the list.

In [ ]:
mix_list.insert(1, 'hallo')
mix_list

You can remove elements by value with *remove*, which locates the first such value and removes from the list.

In [ ]:
mix_list.remove('hola')
mix_list

To illustrate what happens with repeat values, look at the example below.

In [ ]:
rep_list = ['hello', 'hola', 'ciao', 18, 'hola']
rep_list

In [ ]:
rep_list.remove('hola')   # remove only the first 'hola'
rep_list

In [ ]:
rep_list.remove('hola')
rep_list

To concatenate lists, you have two options. You either use the + sign or the *extend* method.

In [ ]:
x = [1, 2, 3]
y = [4, 5, 6]

In [ ]:
x + y

In [ ]:
x.extend(y)
x

In [ ]:
y

Note the difference between the two approaches: x + y creates a new list, which is the concatenation of x and y, but it does not change x nor y. On the other hand, x.extend(y) modifies x by concatenating y to it.

You can sort a list by calling the *sort* method or the *sorted* function. With a list of integers, the sort method works in a straightforward way.

In [ ]:
int_list = [10, 4, 2, 8, 6]

In [ ]:
int_list.sort()
int_list

In [ ]:
int_list = [10, 4, 2, 8, 6]
sorted(int_list)

In [ ]:
help(sorted)

In [ ]:
help(list.sort)

If you want to reverse the list, you can use the *reverse* method. 

In [ ]:
int_list = [10, 4, 2, 8, 6]
int_list.reverse()
int_list

In case you wish to sort a list of strings or a mix of data types, you need to pass a secondary *sort key*, i.e. a function that produces a value used to sort the objects.

In [ ]:
object_list = ['Python', 'statistics', 'love', 'hate']

In [ ]:
object_list.sort()
object_list

In [ ]:
object_list.sort(key = len)
object_list

In [ ]:
object_list.sort(key = str.lower)
object_list

## Tuple

Tuples are essentially lists. Tuples are used to group objects of different types. In contrast to lists, tuples cannot be modified after creating them (i.e. they are *immutable*). 

We can rewrite our first list as a tuple by using round brackets or by invoking *tuple*.

In [54]:
first_tuple = (2, 4, 6, 8, 10)
first_tuple

(2, 4, 6, 8, 10)

In [56]:
type(first_tuple)

tuple

In [58]:
tup = tuple(['hello', 2, 4, True])
tup

('hello', 2, 4, True)

If we try to change the fourth element of tuple, we get an error message:

In [60]:
tup[3] = "False"

TypeError: 'tuple' object does not support item assignment

Indeed, once the tuple is created, it is not possible to modify the object is stored in each slot. 

There are only two methods for tuples of numbers, since the size and content of a tuple cannot be modified. 
The first method is *count*, which counts how many times a specific value is present in the tuple.

In [62]:
num_tup = (2, 4, 6, 8, 8, 8, 10)
num_tup.count(8)

3

The second method is *index*, which tells us at what index the given number first appears in the tuple.

In [65]:
num_tup.index(8)

3

Why/when would you use a tuple instead of a list? 
* Tuples are **faster**. This can become relevant if you need to iterate over a large number of items.
* Tuples are **safer**. As they are immutable, they are basically write-protected, which can be important for a complicated project.
* Tuples can be used as **keys** in dictionary objects (see later).

## Dictionary

Dictionaries or dicts are very important Python data structures. Dictionaries are collections of content in the form of **key-value pairs**. 

You can create dictionaries by using the command *dict*, or by using curly brackets { } in association with the column operator ':' to separate keys and values.

In [67]:
dict_course = dict(organizer = "Flames", topic = "programming", software = "Python", level = "introduction")
dict_course

{'organizer': 'Flames',
 'topic': 'programming',
 'software': 'Python',
 'level': 'introduction'}

In [69]:
course = {'organizer':'Flames', 'topic':'programming', 'software':'Python', 'level':'introduction'}
course

{'organizer': 'Flames',
 'topic': 'programming',
 'software': 'Python',
 'level': 'introduction'}

To access or insert elements inside a dictionary, use square brackets.

In [71]:
course['topic']

'programming'

You can easily add a key-value pair to an existing dictionary.

In [73]:
course['participants'] = 45
course

{'organizer': 'Flames',
 'topic': 'programming',
 'software': 'Python',
 'level': 'introduction',
 'participants': 45}

Use the *del* keyword or the *pop* method to delete a key-value pair.

In [75]:
del course['participants']

In [77]:
course

{'organizer': 'Flames',
 'topic': 'programming',
 'software': 'Python',
 'level': 'introduction'}

In [79]:
course.pop('organizer')

'Flames'

In [81]:
course

{'topic': 'programming', 'software': 'Python', 'level': 'introduction'}

You can extract all keys from the dictionnary using the keys method.

In [83]:
course.keys()

dict_keys(['topic', 'software', 'level'])

The result is stored in an object of type dict_keys.

In [85]:
type(course.keys())

dict_keys

You can convert it to a list using the list function, if needed.

In [87]:
list_course = list(course)
list_course

['topic', 'software', 'level']

In [89]:
type(list_course)

list

Important remark: keys of a dict have to be immutable objects (int, float, str) or tuples. You can check whether an object can be used as a key in a dict with the *hash* function:  

In [ ]:
hash([2, 4, 6]) # error because lists are mutable!

In [93]:
dict_scores = dict(Maths = 78, Economics = 85, Stats = 90, Python = 88, English = 72)
dict_scores

{'Maths': 78, 'Economics': 85, 'Stats': 90, 'Python': 88, 'English': 72}

In [99]:
scores = {" Maths ": 78, " Economics ": 85, " Stats ": 90,
" Python ": 88, " English ": 72}
scores

{' Maths ': 78,
 ' Economics ': 85,
 ' Stats ': 90,
 ' Python ': 88,
 ' English ': 72}


## Exercise: working with lists, tuples, and dictionaries

**Exercise 1**

Start by creating a list with the following elements:

*welcome, yes, 15, sport, 15.02, monday, holiday, 27*

Then:
- Append the value 'wrong' to the list
- Get the second element
- Get the last element
- Get every second element
- Select the first three elements
- Use count and reverse method on the list 
- Remove the value 'wrong'

**Exercise 2**

Start by creating a tuple with the following elements:

 *Leuven, 50.87, 4.70, Belgium, Ghent, 51.05, 3.72, Belgium*

Then:
- Find the index for the value Ghent
- Get the count for the value Belgium    

**Exercise 3**

Create a dict with following personal information:
      - your nationality; 
      - your hobby;
      - your favourite food;
      - if you have siblings: yes or no
      - You like statistics: True or False

**Exercise 4**

Create a dict containing your travel destinations and the reason for the trips. Use a different approach from the one 
    you have used in the previous exercise.
    Example:  Nepal:trekking

**Exercise 5**

Starting from the dict created in exercise **4**, add the year of your trip as additional information.
    Example: 

In [ ]:
trips_year = {}    # alternative way of creating a dict
trips_year['Nepal'] = {'reason':'trekking', 'year':2019}   # continue with the other trips 

#  Conditional and Iterative Processing

Python has several built-in keywords for conditional logic and loops. 

## Conditional Processing: if, elif, and else 

**The *if* Statement**

The *if* statement is one of the most important control flow statement types. We use the *if* statement combined with a condition/statement that returns True of False. If that condition is True, then the code in the block that follows is evaluated. If the condition is False, then the code is not evaluated.

In [101]:
age = 20

if age >= 18:
    print("You are an adult.")

You are an adult.


In [103]:
x = -5
if x < 0:
    print("This is a negative number")

This is a negative number


**The *if-else* Statement**

Optionally, you can provide code that has to be evaluated when the condition is false using *else*.

In [111]:
age = 45

if age >= 18:
    print("You are an adult.")
else:
    print("You are a minor.")

You are an adult.


In [115]:
x = -100
if x < 0:
    print("This is a negative number")
else:
    print("This is a nonnegative number")

This is a negative number


**The *elif* Statement**

Optionally, you can use one or more *elif* blocks, but the final statement has to be *else*. The conditions are evaluated in order. If one of them is True, the subsequent code is evaluated and no further *elif* and *else* blocks will be evaluated.  

In [117]:
x = 50
if x > 100:
    print("this number is big")
elif x > 10:
    print("this number is not big and not small")
else:
    print("this number is small")

this number is not big and not small


In [123]:
score = 60

if score >= 80:
    print("Grade: A")
elif score >= 70:
    print("Grade: B")
elif score >= 60:
    print("Grade: C")
elif score >= 50:
    print("Grade: D")
else:
    print("Grade: F")

Grade: C


## Iterative Processing: for and while 

In chapter 3, we saw how to create, access, and manipulate lists, tuples, and dictionaries. In case we want to perform more than one operation on a collection of data contained in a list for example, we can use loops. 
First we create some patient data (patient IDs, type of cronic disease, cholesterol values):

In [125]:
patient = []
patient1 = ["ID0101", "diabetes", 239]
patient.append(patient1)
patient2 = ["ID0102", "cancer", 201]
patient.append(patient2)
patient3 = ["ID0103", "epilepsy", 171]
patient.append(patient3)
patient4 = ["ID0104", "allergy", 163]
patient.append(patient4)
patient5 = ["ID0105", "hypertension", 252]
patient.append(patient5)
patient

[['ID0101', 'diabetes', 239],
 ['ID0102', 'cancer', 201],
 ['ID0103', 'epilepsy', 171],
 ['ID0104', 'allergy', 163],
 ['ID0105', 'hypertension', 252]]

In [127]:
type(patient)

list

###  *for* loops

Why use loops? If you want to avoid to write the same code over and over again, you might want to use a *for* loop. Below you can find one of the simplest examples of using a *for* loop.

In [129]:
for i in range(5):
    print(i)

0
1
2
3
4


In [131]:
cities = ["New York", "London", "Paris", "Brussels", "Tokyo"]
for c in cities: 
    print(c)

New York
London
Paris
Brussels
Tokyo


In [133]:
fruits = ["apple", "banana", "mango", "orange"]

for f in fruits:
    print(f)

apple
banana
mango
orange


For example, if we want to access the average cholesterol values of all patients in the list that we've just created, we can use the following code: 

In [135]:
tot_chol = 0.0
for p in patient:
    chol = p[2]
    tot_chol += chol
tot_chol

1026.0

In [137]:
avg_chol = tot_chol / len(patient)
avg_chol

205.2

The average cholosterol level in our patient data is 205.2. Thanks to using a *for* loop, we saved time in computing the mean. However, what we really have done here? 

First we create the variable that will contain the sum of all cholesterol levels and we initialize it to zero. Then we use a *for loop*. The syntax says that for every *p* in patient, the cholesterol values (in position 2) are accessed and then the total chol level will be incremented by the individual cholesterol levels within the loop.  

Two remarks about the code:
* Using the letter *p* is arbitrary. You could use any other letter or word. However, using the first letter of the list you are looping is the convention. 

* The operator += is the addition assignment operator in Python. It basically adds the value of the second variable to the first one, and assigns the result to the first variable: A += B is the same as to A = A + B.

###  *while* loops

A *while* loop repeats a block of code **as long as** a condition is true.

It is like a repeated *if* statement. It specifies a condition and a block of code that is executed over and over again as long as the condition is *True*.

In [140]:
count = 5
while count < 9:
    print ('The count is:', count)
    count = count + 1
    
print ("Good bye!")

The count is: 5
The count is: 6
The count is: 7
The count is: 8
Good bye!


In [144]:
count = 1
while count <= 5:
       print (f" Count is { count }")
       count = count + 1
print (" Done !")

 Count is 1
 Count is 2
 Count is 3
 Count is 4
 Count is 5
 Done !


Remember that *while* loops should always end at some point! Pay attention not to create infinite loops!

## Exercise: Conditional and Iterative Processing


**Exercise 1: working with conditional statements**

Write code to generate the appropriate result regarding patients cholesterol levels:
- Total cholesterol levels below 200 milligrams per deciliter (mg/dL) are considered desirable. 
- A reading between 200 and 239 mg/dL is considered borderline high and a reading of 240 mg/dL and above is considered high.

Use a nested *if* statement to obtain a description for a specific cholesterol level. For example: if a patient has a cholesterol level of 190, the resulting description should be "desirable", and so on. 

**Exercise 2: working with for loops**

The variable *house* is a list of lists, where each sublist contains the name and area of a room in a house.

In [ ]:
house = [["hallway", 11.25], 
         ["kitchen", 18.0], 
         ["living room", 20.0], 
         ["bedroom", 10.75], 
         ["bathroom", 9.50]]

Write a *for* loop that goes through each sublist of house and prints out "x is y sqm", where *x* is the name of the room and *y* is the area of the room.

**Exercise 3: working with while loops**

In a series of experiments, you initially have to deal with an error of 60.0. You want to divide this error by 4 on every run and to continue until the error is no longer above 1. By using a *while* loop, try to get an error lower than 1.